plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs

```diff
@@ -1578,63 +1591,56 @@ checkNoLockedFundsProof
+checkNoLockedFundsProof options =
+  checkNoLockedFundsProof' prop
   where
     prop = propRunActionsWithOptions' options defaultCoverageOptions (\ _ -> pure True)
 
-checkNoLockedFundsProofWithWiggleRoomFast
+checkNoLockedFundsProofFast
   :: (ContractModel model)
-  => Integer
+  => CheckOptions
   -> NoLockedFundsProof model
   -> Property
-checkNoLockedFundsProofWithWiggleRoomFast wiggle = checkNoLockedFundsProofWithWiggleRoom' wiggle (const $ property True)
+checkNoLockedFundsProofFast _ = checkNoLockedFundsProof' (const $ property True)
 
-checkNoLockedFundsProofWithWiggleRoom'
+checkNoLockedFundsProof'
   :: (ContractModel model)
-  => Integer
-  -> (StateModel.Actions (ModelState model) -> Property)
+  => (StateModel.Actions (ModelState model) -> Property)
   -> NoLockedFundsProof model
   -> Property
-checkNoLockedFundsProofWithWiggleRoom' wiggle run NoLockedFundsProof{nlfpMainStrategy   = mainStrat,
-                                                                     nlfpWalletStrategy = walletStrat } =
+checkNoLockedFundsProof' run NoLockedFundsProof{nlfpMainStrategy   = mainStrat,
+                                                nlfpWalletStrategy = walletStrat,
+                                                nlfpOverhead       = overhead,
+                                                nlfpErrorMargin    = wiggle } =
     forAllDL anyActions_ $ \ (Actions as) ->
-    forAllUniqueDL (nextVarIdx as) (stateAfter $ Actions as) mainStrat $ \ (Actions as') ->
+    let s0 = (stateAfter $ Actions as) in
+    forAllUniqueDL (nextVarIdx as) s0 mainStrat $ \ (Actions as') ->
           let s = stateAfter $ Actions (as ++ as') in
-            foldl (QC..&&.) (counterexample "Main run prop" (run (toStateModelActions $ Actions $ as ++ as')) QC..&&. (counterexample "Main strategy" . counterexample (show . Actions $ as ++ as') $ prop s))
-                            [ walletProp as w bal | (w, bal) <- Map.toList (s ^. balanceChanges) ]
+            foldl (QC..&&.) (counterexample "Main run prop" (run (toStateModelActions $ Actions $ as ++ as')) QC..&&. (counterexample "Main strategy" . counterexample (show . Actions $ as ++ as') $ prop s0 s))
+                            [ walletProp s0 as w bal | (w, bal) <- Map.toList (s ^. balanceChanges) ]
     where
         nextVarIdx as = 1 + maximum ([0] ++ [ i | Var i <- varOf <$> as ])
-        prop s =
+        prop s0 s =
           let lockedVal = lockedValue s
-          in (counterexample ("Locked funds should be zero, but they are\n  " ++ show lockedVal) $ symIsZero lockedVal)
+          in (counterexample ("Locked funds should be at most " ++ show (overhead s0) ++ ", but they are\n  " ++ show lockedVal)
+            $ symLeq lockedVal (overhead s0))
 
-        walletProp as w bal = DL.forAllUniqueDL (nextVarIdx as) (stateAfter $ Actions as) (DL.action (Unilateral w) >> walletStrat w) $ \ acts ->
+        walletProp s0 as w bal =
+          let wig = wiggle s0
```

Comment:
Is this code so slow it needs to sit outside of the `DL.forAllUniqueDL` loop? 
(It looks a bit like a hack in its current form.)

Reply:
The problem is that it's user code - so we don't know how fast or slow it is. Therefore we try to limit the number of footguns. On the other hand - it's only used twice when reporting errors so I don't think performance here is a super big deal.